elephant/spike_train_surrogates.py

```diff
@@ -147,13 +143,95 @@ def dither_spikes(spiketrain, dither, n=1, decimals=None, edges=True):
 
 
+def dither_spikes_with_refractory_period(spiketrain, dither, n_surrogates=1,
+                                         refractory_period=4*pq.ms):
+    """
+    Generates surrogates of a spike train by uniform dithering conserving the
+    refractory period.
+
+    The surrogates are obtained by uniformly dithering times around the
+    original position. The dithering is performed independently for each
+    surrogate. The dithering range is for each spike adjusted such, that it
+    conserves the refractory period.
+
+    The surrogates retain the :attr:`t_start` and :attr:`t_stop` of the
+    original `SpikeTrain` object.
+
+    Parameters
+    ----------
+    spiketrain : neo.SpikeTrain
+        The spiketrain from which to generate the surrogates from.
+    dither : pq.Quantity
+        Amount of dithering.
+        To a spike a uniform random step within the window
+        '[-previous_dither, next_dither]' is added. With:
+        'previous_dither = min(dither,
+            spike - previous_spike - refractory_period)'
+        'next_dither = min(dither,
+            next_spike - spike - refractory_period)'
+    n_surrogates : int, optional
+        Number of surrogates to create.
+        Default: 1
+    refractory_period
+        The starting 'refractory_period' to work with.
+        If 'min(isi) < refractory_period', function works with 'min(isi)'
+        as 'refractory_period'.
+        Default: 4.*pq.ms
+
+    Returns
+    -------
+    dithered_spiketrains : list of neo.SpikeTrain
+        Each element is an independent surrogate spiketrain of 'spiketrain',
+        where each spike is moved.
+    """
+    if len(spiketrain) == 0:
+        return [spiketrain.copy() for _ in range(n_surrogates)]
+
+    units = spiketrain.units
+    t_start, t_stop = spiketrain.t_start.magnitude, spiketrain.t_stop.magnitude
+
+    dither = dither.rescale(units).magnitude
+    refractory_period = refractory_period.rescale(units).magnitude
+    refractory_period = np.min(np.diff(spiketrain.magnitude),
+                               initial=refractory_period)
+
+    dithered_spiketrains = []
+    for _ in range(n_surrogates):
+        dithered_spiketrain = np.copy(spiketrain.magnitude)
+        random_ordered_ids = np.arange(len(spiketrain))
+        np.random.shuffle(random_ordered_ids)
+
+        for random_id in random_ordered_ids:
+            spike = dithered_spiketrain[random_id]
+
+            previous_spike = dithered_spiketrain[random_id - 1] \
+                if random_id > 0 \
+                else t_start - refractory_period
+            next_spike = dithered_spiketrain[random_id + 1] \
+                if random_id < len(spiketrain) - 1 \
+                else t_stop + refractory_period
+
+            previous_dither = min(dither,
+                                  spike - previous_spike - refractory_period)
+            next_dither = min(dither,
+                              next_spike - spike - refractory_period)
```

Comment:
are you sure it should not be like 
`min(dither, next_spike - spike + refractory_perioud)`?

Reply:
Yes, I'm sure. This keeps the distance between the spike and the next_spike, to be greater than the refractory period.